README
.. $Id:$    -*- rst -*-

..
  This file is part of pyFormex 0.9.1  (Tue Oct 15 21:05:25 CEST 2013)
  pyFormex is a tool for generating, manipulating and transforming 3D
  geometrical models by sequences of mathematical operations.
  Home page: http://pyformex.org
  Project page:  http://savannah.nongnu.org/projects/pyformex/
  Copyright 2004-2013 (C) Benedict Verhegghe (benedict.verhegghe@ugent.be)
  Distributed under the GNU General Public License version 3 or later.

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see http://www.gnu.org/licenses/.


What is pyFormex?
=================
pyFormex is a tool for generating, manipulating and transforming large
geometrical models of 3D structures by sequences of mathematical
transformations. Thanks to a powerful (Python based) scripting language,
pyFormex is very well suited for the automated design of spatial frame
structures. It provides a wide range of operations on surface meshes,
like STL type triangulated surfaces. There are provisions to import medical
scan images. pyFormex can also be used as a pre- and post-processor for
Finite Element analysis programs. Finally, it might be used just for
creating some nice graphics.

Using pyFormex, the topology of the elements and the final geometrical form
can be decoupled. Often, topology is created first and then mapped onto the
geometry. Through the scripting language, the user can define any sequence
of transformations, built from provided or user defined functions.
This way, building parametric models becomes a natural thing.

While pyFormex is still under development (see the project page at
http://savannah.nongnu.org/projects/pyformex/), it already provides a
fairly stable scripting language and an OpenGL GUI environment for
displaying and manipulating the generated structures.


Installation
============

We detail here only the installation on Linux platforms.


Prerequisites
-------------
Essential:

- python: http://www.python.org  (version 2.4 or higher; 2.5 recommended)
- numpy: /http://numpy.scipy.org/ (version 1.0 or higher; 1.1 recommended)
- Qt4: http://www.trolltech.com/products/qt
- PyQt4: http://www.riverbankcomputing.co.uk/pyqt/index.php
- PyOpenGL: http://pyopengl.sourceforge.net/

To compile the acceleration library (highly recommended!), you will also
need the appropriate header Python and OpenGL header files.

Further, we recommend to install the following for extended functionality:

- python-gnuplot
- python-doc
- python-scipy
- units
- imagemagick
- admesh

For the following extensions, which are not easily packaged format,
an install script is provided in this distribution.

- calpy
- pygl2ps
- gts
- tetgen


Easy way to install prerequisites on Debian GNU/Linux or Ubuntu
---------------------------------------------------------------
On Debian (lenny) systems (and Debian-derivatives like Ubuntu) you can install
all basic prerequisites and their dependencies with the command::

  apt-get install python-qt4-gl python-numpy

If you want to compile the acceleration library (which is highly recommended) you also need the header files::

  apt-get install python-dev libgl1-mesa-dev

Install the extras::

  apt-get install python-gnuplot python-doc python-scipy units imagemagick admesh

Howto install pyFormex:
-----------------------
If a previous version of pyFormex was already installed on your system,
you may want to remove the older version first before installing the new one.
See further (Howto uninstall pyFormex)

We suppose you have installed all the prerequisites and downloaded the
pyformex source tarball: pyformex-VERSION.tar.gz

Unpack the pyFormex source tarball::

  tar xvzf pyformex-VERSION.tar.gz

Go to the created pyformex directory::

  cd pyformex-VERSION

Do (with root privileges) ::

  python setup.py install --prefix=/usr/local

This will install pyFormex under /usr/local/. You can change the prefix
to install in some other place.

After installation you can remove the directory where you unpacked the tarball.

The installation procedure installs everything into a single directory,
and creates a symlink to the executable in /usr/local/bin.
You can use the command pyformex --whereami to find out where pyFormex
is installed.

If you have xdg-utils on your system, the installation procedure will also
install a menu and desktop starter for pyFormex.


Howto uninstall pyFormex:
-------------------------
A pyFormex installation of version 0.5 or later can be removed with the command ::

  pyformex --remove

and then answering 'yes' to the question.



Howto install the extra packages
-----------------------------------
pyFormex makes use of extra software components to enhance its functionality.
While they are not required for the core operation of pyFormex, many users may
want to install them. Some of these extras (admesh, units) can easily be
installed from your regular distribution repositories.
Some extra components however either are not available in packaged format, or
the existing packages do not work together well with pyFormex.
For these components, the pyFormex distribution contains an adhoc install
procedure to help our users with the installation. The procedures are located
in dedicated subdirectories of the pyformex-VERSION/pyformex/extra directory
of the unpacked release tree. Each subdirectory contains a ``README`` file
with installation instructions. It the directory contains a ``Makefile``, this
usually boils down to the usual::

  make
  make install

Else, an install script ``install.sh`` is included and installation can be done
with by executing it from within the subdirectory, with root privileges and with a single parameter 'all'::

  (sudo) ./install.sh all

- gts: Installs some programs to do operations on triangulated surfaces. You need to have the ``libgts`` and its header files installed. On Debian::

    apt-get install libgts-dev

- dxfparser: Installs the command ``pyformex-dxfparser``, to preprocesses AutoCAD .DXF
  files for import into pyFormex.
- postabq: Installs the command ``pyformex-postabq``, to preprocess Abaqus output files (.fil) for import into pyFormex.
- gl2ps: Image output to in vector formats (PS, EPS, PDF and SVG). Requires
  a Python interface, provided by our install procedure.
- calpy: Simple finite element simulation framework. Needed for the ``*_calpy``
  examples. Has to be installed from source, because is has to be compiled
  against the same numpy version as pyFormex itself.


Running pyFormex from Subversion (SVN) sources
----------------------------------------------
If the officially released pyFormex packages are not suitable for your needs,
you can try installing pyFormex from the Subversion sources. Besides the basic
prerequisites (see above), you will also need to have 'Subversion' installed.

The `SVN repository`_ on the developer site is accessible by anonymous SVN.
The command ::

  svn co svn://svn.savannah.nongnu.org/pyformex/trunk/pyformex MYDIR

will checkout the source to a local directory ``MYDIR``. Provided you have all the prerequisites installed, pyFormex can then be run directly from the checked out source with the command::

  MYDIR/pyformex/pyformex

If you want to use the compiled accelerator library however, you will have to
create it first::

  cd MYDIR
  make lib

Once you have a checked-out source tree, you can easily sync it to the latest
repository version by just issuing the following command from your checkout
directory::

  svn up

Documentation
=============

The documentation is included with the pyFormex distribution. Execute the command
pyformex --whereami to find out where the pyformex files are
installed. The pyformex path contains a directory doc/html with the
full html ocumentation.
When working with the pyFormex GUI, you can load this local documentation in your browser
using the help menu, or view the online documentation at
http://pyformex.org/doc.

To use pyFormex, you create a script that builds a structure layout and
geometry. Look at the examples to get the feeling of how it is working.
Use the File->Play menu selection to display the structure.

The examples and the Python source are relatively well documented.


Help and Feedback
=================

For any questions or feedback, please go to the pyFormex support tracker at
http://savannah.nongnu.org/support/?group=pyformex

If you find any bugs or malfunctions in pyFormex, please submit them to
the pyFormex Bug tracker http://savannah.nongnu.org/bugs/?group=pyformex

.. _`SVN repository`: http://savannah.nongnu.org/svn/?group=pyformex

.. target-notes::


Copyright
=========

pyFormex is distributed under the GNU GPL version 3 or later.

The main author claims that he wrote this software in his free time,
for his joy, not as a commissioned task. Any copyright claims made by
his employer (previous, current r future) should therefore be considered
void.

Over the years, other members of the pyFormex team have made valuable
contributions. These were made with full consent of the contributor that
they could be used at will inside the pyFormex project and were free of
other copyright claims.

.. End


History
# $Id:$
##
##  This file is part of the pyFormex project.
##  pyFormex is a tool for generating, manipulating and transforming 3D
##  geometrical models by sequences of mathematical operations.
##  Home page: http://pyformex.org
##  Project page:  https://savannah.nongnu.org/projects/pyformex/
##  Copyright (C) Benedict Verhegghe (benedict.verhegghe@ugent.be) 
##  Distributed under the GNU General Public License version 3 or later.
##
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see http://www.gnu.org/licenses/.
##

2009-06 BV: Released pyFormex 0.8

2008-04 BV: Released pyFormex 0.7

2008-02 BV: Started the pyFormex User Meeting

2007-11 BV: Released pyFormex 0.6

2007-08 BV: Released pyFormex 0.5, a major step towards a multifunctional
            program.

2007-02 BV: Released pyFormex 0.4.2, including lots of work on surface
            models.

2006-11-03 BV: Released pyFormex 0.4. This release is marked by the transition
               to the Qt4 widget set.

2006-04-15 BV: Released pyFormex 0.3

2006: pyFormex was used in some of our research projects and proved to be
      a valuable tool. This necessitated and inspired new development work
      on pyFormex. Lots of new features were implemented, too many to be
      listed here.
      Some of the milestones:
      - pyFormex was used in the design of wire stents (see WireStent example).
        The process involved the automatic creation of Finite Element
        Analysis jobs and automatic processing of the results.
      - pyFormex was used to measure geometries obtained from medical images
        and to create a patient specific geometrical hardware model. This
        initiated our further work with surface models.
      - pyFormex was used by civil engineering students for their projects
        in designing shell structures. Some of these will be included as
        examples in future releases. 


2005-12-09 BV : First anniversary release 0.2.2
2005-04-23 BV : Interim 0.2.1 release with lots of new stuff
2005-01-03 BV : First public release on berlios.de (0.2)
2004-12-09 BV : Registered project pyFormex on berlios.de
2004-07-08 BV : Created a first release (0.1.1)
2004-07-02 BV : Created a Subversion repository
2004-06-29 BV :	After reading about Formian, I quickly hacked together
                the Formex class

# End

manifest.py
#!/usr/bin/env python
# $Id$
##
##  This file is part of pyFormex 0.9.1  (Tue Oct 15 21:05:25 CEST 2013)
##  pyFormex is a tool for generating, manipulating and transforming 3D
##  geometrical models by sequences of mathematical operations.
##  Home page: http://pyformex.org
##  Project page:  http://savannah.nongnu.org/projects/pyformex/
##  Copyright 2004-2013 (C) Benedict Verhegghe (benedict.verhegghe@ugent.be)
##  Distributed under the GNU General Public License version 3 or later.
##
##  This program is free software: you can redistribute it and/or modify
##  it under the terms of the GNU General Public License as published by
##  the Free Software Foundation, either version 3 of the License, or
##  (at your option) any later version.
##
##  This program is distributed in the hope that it will be useful,
##  but WITHOUT ANY WARRANTY; without even the implied warranty of
##  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
##  GNU General Public License for more details.
##
##  You should have received a copy of the GNU General Public License
##  along with this program.  If not, see http://www.gnu.org/licenses/.
##
#
"""manifest.py

This script creates the list of files to be included in
the pyFormex source distribution.
"""
import os,re

def prefixFiles(prefix,files):
    """Prepend a prefix to a list of filenames."""
    return [ os.path.join(prefix,f) for f in files ]

def matchMany(regexps,target):
    """Return multiple regular expression matches of the same target string."""
    return [re.match(r,target) for r in regexps]


def matchCount(regexps,target):
    """Return the number of matches of target to  regexps."""
    return len(filter(None,matchMany(regexps,target)))


def matchAny(regexps,target):
    """Check whether target matches any of the regular expressions."""
    return matchCount(regexps,target) > 0


def matchNone(regexps,target):
    """Check whether targes matches none of the regular expressions."""
    return matchCount(regexps,target) == 0


def listTree(path,listdirs=True,topdown=True,sorted=False,excludedirs=[],excludefiles=[],includedirs=[],includefiles=[],filtr=None):
    """List all files in path.

    If ``dirs==False``, directories are not listed.
    By default the tree is listed top down and entries in the same directory
    are unsorted.

    `exludedirs` and `excludefiles` are lists of regular expressions with
    dirnames, resp. filenames to exclude from the result.

    `includedirs` and `includefiles` can be given to include only the
    directories, resp. files matching any of those patterns.

    Note that 'excludedirs' and 'includedirs' force top down handling.
    """
    filelist = []
    if excludedirs or includedirs:
        topdown = True
    for root, dirs, files in os.walk(path, topdown=topdown):
        if sorted:
            dirs.sort()
            files.sort()
        if excludedirs:
            remove = [ d for d in dirs if matchAny(excludedirs,d) ]
            for d in remove:
                dirs.remove(d)
        if includedirs:
            remove = [ d for d in dirs if not matchAny(includedirs,d) ]
            for d in remove:
                dirs.remove(d)
        if listdirs and topdown:
            filelist.append(root)
        if excludefiles:
            files = [ f for f in files if not matchAny(excludefiles,f) ]
        if includefiles:
            files = [ f for f in files if matchAny(includefiles,f) ]
        filelist.extend(prefixFiles(root,files))
        if listdirs and not topdown:
            filelist.append(root)
    if filtr:
        filelist = [ f for f in filelist if filtr(f) ]
    return filelist


#NON_MANIFEST = [ f.strip('\n') for f in open('NONMANIFEST').readlines() ]


# pyFormex documentation (installed in the pyformex tree)
DOC_FILES = listTree(
    'pyformex/doc/html',listdirs=False,sorted=True,
    excludedirs=['.svn'],
    ) + listTree(
    'pyformex/doc',
    listdirs=False,sorted=True,
    excludedirs=['.svn','dutch','html'],
    includefiles=[
        'README',
        'Description',
        'COPYING',
        'ReleaseNotes',
        ],
    )


# pyFormex data files (installed in the pyformex tree)
DATA_FILES = listTree(
    'pyformex/data',listdirs=False,sorted=True,
    excludedirs=['.svn','benchmark','ply'],
    excludefiles=['.*\.pyc','.*~$','PTAPE.*'],
    includefiles=[
        'README',
        'benedict_6\.jpg',
        'bifurcation\.off',
        'blippo\.pgf',
        'blippok\.ttf',
        'butterfly\.png',
        'fewgl\.js',
        'fragment_shader.*\.c',
        'hesperia-nieve\.prop',
        'horse\.off',
        'horse\.pgf',
        'materials\.db',
        'sections\.db',
        'splines\.pgf',
        'supershape\.txt',
        'teapot\.off',
        'vertex_shader.*\.c',
        'world\.jpg',
        'xtk_xdat\.gui\.js',
        ],
    )


# scripts to install extra programs
EXTRA_FILES = listTree(
    'pyformex/extra',listdirs=True,sorted=True,
    includedirs=[
        'calpy',
        'dxfparser',
###        'freetype-py',  now included in pyformex
        'gts',
        'postabq',
        'pygl2ps',
        'tetgen',
        ],
    excludefiles=['.*~$','.*\.1\.rst'],
    includefiles=[
        'README',
        'Makefile',
        '.*\.sh$',
        '.*\.rst$',
        '.*\.patch$',
        '.*\.c$',
        '.*\.cc$',
        '.*\.h$',
        '.*\.i$',
        '.*\.py$',
        '.*\.1$',
        '.*\.pc$',
#        'freetype-py-0.4.1.tar.gz',
        ],
    )

# Data files to be installed outside the pyformex tree
# These are tuples (installdir,filelist)
OTHER_DATA = [
    ('share/pixmaps', [
        'pyformex/icons/pyformex-64x64.png',
        'pyformex/icons/pyformex.xpm',
        ]),
    ('share/applications', ['pyformex.desktop']),
    ('share/man/man1', [
        'pyformex/doc/pyformex.1',
        ]),
    # the full html documentation
    ## ('share/doc/pyformex/html',DOC_FILES),
    ]

DIST_FILES = [
    'README',
    'COPYING',
    'Description',
    'ReleaseNotes',
    'manifest.py',
    'setup.py',
    'setup.cfg',
    ] + \
    listTree('pyformex',listdirs=False,sorted=True,
             includedirs=['gui','plugins','opengl','legacy'],
             includefiles=['.*\.py$','pyformex(rc)?$','pyformex.conf$'],
             excludefiles=['core.py','curvetools.py','backports.py', 'compat_3k.py'],
             ) + \
    listTree('pyformex/freetype',listdirs=False,sorted=True,
             includedirs=['ft_enums'],
             includefiles=['.*\.py$'],
             ) + \
    listTree('pyformex/icons',listdirs=False,sorted=True,
             includefiles=['README','.*\.xpm$','.*\.png$']
             ) + \
    listTree('pyformex/lib',listdirs=False,sorted=True,
             includefiles=['.*\.c$','.*\.py$']
             ) + \
    listTree('pyformex/bin',listdirs=False,sorted=True,
             excludefiles=['.*~$'],
             ) + \
    listTree('pyformex/examples',listdirs=False,sorted=True,
             excludedirs=['Demos'],
             excludefiles=['.*\.pyc','.*~$',
                           ],
             includefiles=['[_A-Z].*\.py$','apps.cat','README']
             ) + \
    DATA_FILES + \
    DOC_FILES + \
    EXTRA_FILES


for i in OTHER_DATA:
   DIST_FILES += i[1]

if __name__ == '__main__':
   import sys

   todo = sys.argv[1:]
   if not todo:
      todo = ['doc','data','dist']

   for a in todo:
      if a == 'doc':
         print "=========DOC_FILES========="
         print '\n'.join(DOC_FILES)
      elif a == 'data':
         print "=========DATA_FILES========"
         print '\n'.join(DATA_FILES)
      elif a == 'other':
         print "=========OTHER_DATA========"
         for i in OTHER_DATA:
            print '\n'.join(i[1])
      else:
         print "=========DIST_FILES========"
         print '\n'.join(DIST_FILES)

# End
